pkg/reconciler/stats_reporter.go

```diff
@@ -0,0 +1,172 @@
+
+package reconciler
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"knative.dev/pkg/metrics"
+)
+
+const (
+	InstallSuccessCountName    = "install_success_count"
+	InstallSuccessLatencyName  = "install_success_latency"
+	DeploymentReadyCountName   = "deployment_ready_count"
+	DeploymentReadyLatencyName = "deployment_ready_latency"
+)
+
+var (
+	InstallSuccessCountStat = stats.Int64(
+		InstallSuccessCountName,
+		"Number of successful KnativeServing install",
+		stats.UnitDimensionless)
+	InstallSuccessLatencyStat = stats.Int64(
+		InstallSuccessLatencyName,
+		"Latency of successful KnativeServing install in milliseconds",
+		stats.UnitMilliseconds)
+	DeploymentReadyCountStat = stats.Int64(
+		DeploymentReadyCountName,
+		"Number of successful knative serving deployment reconcile",
+		stats.UnitDimensionless)
+	DeploymentReadyLatencyStat = stats.Int64(
+		DeploymentReadyLatencyName,
+		"Latency of successful knative serving deployment reconcile in milliseconds",
+		stats.UnitMilliseconds)
+	// Create the tag keys that will be used to add tags to our measurements.
+	// Tag keys must conform to the restrictions described in
+	// go.opencensus.io/tag/validate.go. Currently those restrictions are:
+	// - length between 1 and 255 inclusive
+	// - characters are printable US-ASCII
+	reconcilerTagKey = tag.MustNewKey("reconciler")
+	keyTagKey        = tag.MustNewKey("key")
+)
+
+func init() {
+	// Create views to see our measurements. This can return an error if
+	// a previously-registered view has the same name with a different value.
+	// View name defaults to the measure name if unspecified.
+	if err := view.Register(
+		&view.View{
+			Description: InstallSuccessCountStat.Description(),
+			Measure:     InstallSuccessCountStat,
+			Aggregation: view.Count(),
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+		&view.View{
+			Description: InstallSuccessLatencyStat.Description(),
+			Measure:     InstallSuccessLatencyStat,
+			Aggregation: view.Distribution(10, 100, 1000, 10000, 30000, 60000), // Bucket boundaries are 10ms, 100ms, 1s, 10s, 30s and 60s.
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+		&view.View{
+			Description: DeploymentReadyCountStat.Description(),
+			Measure:     DeploymentReadyCountStat,
+			Aggregation: view.Count(),
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+		&view.View{
+			Description: DeploymentReadyLatencyStat.Description(),
+			Measure:     DeploymentReadyLatencyStat,
+			Aggregation: view.Distribution(10, 100, 1000, 10000, 30000, 60000), // Bucket boundaries are 10ms, 100ms, 1s, 10s, 30s and 60s.
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+	); err != nil {
+		panic(err)
+	}
+}
+
+// StatsReporter defines the interface for sending metrics
+type StatsReporter interface {
+	// ReportKnativeServingReady reports the count and latency metrics for a reconcile operation
```

Comment:
These metrics seem like they are redundant. Should an install be considered successful before the deployments are ready?